### PR TITLE
fix(rolldown): return semantic error from oxc

### LINKED ISSUE #1494: Emit semantic error from oxc
The semantic builder also returns JavaScript syntax errors.

https://github.com/rolldown/rolldown/blob/87ec563e8bc1a2241d80a40452c8dc7c6c59ad06/crates/rolldown_oxc_utils/src/oxc_ast/helpers.rs#L14-L21

https://github.com/oxc-project/oxc/blob/54c653ebd44e0d5968dfac0a983c7f152dede732/crates/oxc_semantic/src/builder.rs#L87

You may abort the program if any error occours.

### PR DESCRIPTION
**Pull request by Boshen**
closes #1494
closes #1910

### CODE REVIEW

**Comment by Boshen** [comment]
```
failures:
    test_tests__esbuild__dce__inline_empty_function_calls__config_json
    test_tests__esbuild__dce__inline_identity_function_calls__config_json
    test_tests__esbuild__dce__multiple_declaration_tree_shaking__config_json
    test_tests__esbuild__dce__multiple_declaration_tree_shaking_minify_syntax__config_json
    test_tests__esbuild__default__keep_names_all_forms__config_json
    test_tests__esbuild__ts__ts_declare_const__config_json
    test_tests__esbuild__ts__ts_declare_const_enum__config_json
    test_tests__esbuild__ts__ts_declare_enum__config_json
    test_tests__esbuild__ts__ts_declare_let__config_json
    test_tests__esbuild__ts__ts_declare_var__config_json
```

It currently prints 

```
---- test_tests__esbuild__ts__ts_declare_var__config_json stdout ----
Input: /Users/boshen/github/rolldown/crates/rolldown/tests/esbuild/ts/ts_declare_var/_config.json
thread 'test_tests__esbuild__ts__ts_declare_var__config_json' panicked at crates/rolldown_testing/src/integration_test.rs:87:5:
Expected the bundling to be success, but got diagnosable errors: [BuildDiagnostic { inner: ParseError { source: "declare var require: any\ndeclare var exports: any;\ndeclare var module: any\n\ndeclare var foo: any\nlet foo = bar()", filename: "/Users/boshen/github/rolldown/crates/rolldown/tests/esbuild/ts/ts_declare_var/entry.ts", error_help: "", error_message: "Identifier `foo` has already been declared", error_labels: [LabeledSpan { label: Some("`foo` has already been declared here"), span: SourceSpan { offset: SourceOffset(88), length: 8 }, primary: false }, LabeledSpan { label: Some("It can not be redeclared here"), span: SourceSpan { offset: SourceOffset(101), length: 3 }, primary: false }] }, source: None, severity: Error }]
```

This is hard to debug, are we able to print some code frames for these?

**Comment by IWANABETHATGUY** [comment]
you could add field: 

```json
{
  ... snip
  "expectError": true
}
```
in `_config.json`

**Comment by IWANABETHATGUY** [comment]
Also, most of them in esbuild couild be bundled, if we emit the error, the bundle would fail. I recommended change the error to warnings.

File: crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
 ---
+# warnings
+
+## PARSE_ERROR
+
+```text
+[PARSE_ERROR] Warning: Identifier `DROP` has already been declared
+   ╭─[identity-last.js:1:10]
+   │
+ 1 │ function DROP(x) { return [x] }
```
Comment (Boshen) [inline crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap:12]: Why do I feel like all these are script mode code.

**Comment by IWANABETHATGUY** [comment]
I made some tweak, considering all semantic diagnostic as warning, since esbuild treat don't even considering these issues as errors.

File: crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap [hunk 2]
```diff
@@ -1,6 +1,23 @@
...
+## PARSE_ERROR
+
+```text
+[PARSE_ERROR] Warning: Identifier `foo` has already been declared
+   ╭─[entry.ts:5:15]
+   │
+ 5 │ declare const foo: any
+   │               ────┬───  
+   │                   ╰───── `foo` has already been declared here
+ 6 │ let foo = bar()
```
Comment (Boshen) [inline crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap:15]: What's the intention of these tests? It feels like we should patch these incorrect tests?

File: crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap:12]: Yeah, but sicne esbuild could emit bundle, that's the only way to align with them. https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.js%00function+DROP%28x%29+%7B+return+%5Bx%5D+%7D%0Afunction+DROP%28x%29+%7B+return+x+%7D%0Aconsole.log%28DROP%281%29%29%0ADROP%28foo%28%29%29%0ADROP%281%29%0A&b=%00file.js%00&o=%7B%0A++treeShaking%3A+true%2C%0A++external%3A+%5B%22c%22%2C+%22a%22%2C+%22b%22%5D%2C%0A%22bundle%22%3A+true%2C%0Aformat%3A+%22esm%22%0A%7D

File: crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap (hunk 2 above)
Comment (hyf0) [reply crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap:15]: https://github.com/rolldown/rolldown/blob/e47e8611a818d28c300160d811907c2306bcb5a4/crates/rolldown/src/utils/parse_to_ecma_ast.rs#L92-L100

https://github.com/rolldown/rolldown/blob/e47e8611a818d28c300160d811907c2306bcb5a4/crates/rolldown/src/utils/parse_to_ecma_ast.rs#L20-L28

Base on the code, they are parsed as module.

File: crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap (hunk 1 above)
Comment (hyf0) [reply crates/rolldown/tests/esbuild/dce/inline_identity_function_calls/artifacts.snap:12]: https://github.com/rolldown/rolldown/blob/e47e8611a818d28c300160d811907c2306bcb5a4/crates/rolldown/src/utils/parse_to_ecma_ast.rs#L92-L100

https://github.com/rolldown/rolldown/blob/e47e8611a818d28c300160d811907c2306bcb5a4/crates/rolldown/src/utils/parse_to_ecma_ast.rs#L20-L28

Base on the code, they are parsed as module.

File: crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap (hunk 2 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap:15]: All snapshots that esbuild could generate output we treat it as correct, esbuild test are treated as incorrect(the input has syntax error and so on), and will not emit any snapshots.
We just align with that behavior.
bundler has better tolerance than a normal parser, we should try to recover (ignore) trivial errors and bundle something to user.

Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/ts/ts_declare_const/artifacts.snap:15]: > What's the intention of these tests? It feels like we should patch these incorrect tests?

There is no very clear incorrect definition in the bundler, so I prefer not to modify the tests.

**Comment by IWANABETHATGUY** [comment]
LGTM, let's wait #2501

**Comment by hyf0** [comment]
Let me solve conflicts caused by https://github.com/rolldown/rolldown/pull/2501.

**Review by IWANABETHATGUY — APPROVED** [review]

**Pull request merged**